### PR TITLE
`azurerm_public_ip` - support for the `domain_name_label_scope` property

### LINKED ISSUE #27730: azurerm_public_ip - Support setting domain label scope
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

Would like a new optional paramteer which can set the  "domain name label scope" on public IP address resources to avoid risk related to dangling DNS names.

Supported values are

```
"NoReuse"
"ResourceGroupReuse"
"SubscriptionReuse"
"TenantReuse"
```
This must be set upon creation of the resource

The API has supported setting this since provider version "Microsoft.Network/publicIPAddresses@2022-11-01", the feature is currently in public preview.

### New or Affected Resource(s)/Data Source(s)

azurerm_public_ip

### Potential Terraform Configuration

```hcl
resource "azurerm_public_ip" "example" {
  name                = "acceptanceTestPublicIp1"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  allocation_method   = "Static"
  domain_label_scope = "TenantReuse" # The domain name label scope can only be specified at the creation of a public IP address
}
```

### References

https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/public-ip-addresses?branch=pr-en-us-268827#domain-name-label-scope-preview
https://learn.microsoft.com/en-us/azure/templates/microsoft.network/publicipaddresses?pivots=deployment-language-terraform

### PR DESCRIPTION
**Pull request by sinbai**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Support for the `domain_name_label_scope` property.

Swagger: https://github.com/Azure/azure-rest-api-specs/blob/fe44d3261ff0ea816315126120672ccec78c3074/specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/publicIpAddress.json#L619

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

PASS: TestAccPublicIpStatic_update (300.85s)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_public_ip` - support for the `domain_name_label_scope` property [GH-27730]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #27730

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
test failure
```
------- Stdout: -------
=== RUN   TestAccPublicIpStatic_canLabelBe63
=== PAUSE TestAccPublicIpStatic_canLabelBe63
=== CONT  TestAccPublicIpStatic_canLabelBe63
    testcase.go:173: Step 1/3 error: Error running pre-apply plan: exit status 1
        
        Error: domain_name_label must contain only lowercase alphanumeric characters, numbers and hyphens. It must start with a letter, end only with a number or letter and not exceed 63 characters in length
        
          with azurerm_public_ip.test,
          on terraform_plugin_test.tf line 43, in resource "azurerm_public_ip" "test":
          43:   domain_name_label = "6lv2zfdpd98dl0yse8w2ajdh0qm8xypfdtbmncy7hjtetbquacbp2ozjeq37by2"
        
--- FAIL: TestAccPublicIpStatic_canLabelBe63 (6.83s)
FAIL
```

**Comment by sinbai** [comment]
> TestAccPublicIpStatic_canLabelBe63



> test failure
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccPublicIpStatic_canLabelBe63
> === PAUSE TestAccPublicIpStatic_canLabelBe63
> === CONT  TestAccPublicIpStatic_canLabelBe63
>     testcase.go:173: Step 1/3 error: Error running pre-apply plan: exit status 1
>         
>         Error: domain_name_label must contain only lowercase alphanumeric characters, numbers and hyphens. It must start with a letter, end only with a number or letter and not exceed 63 characters in length
>         
>           with azurerm_public_ip.test,
>           on terraform_plugin_test.tf line 43, in resource "azurerm_public_ip" "test":
>           43:   domain_name_label = "6lv2zfdpd98dl0yse8w2ajdh0qm8xypfdtbmncy7hjtetbquacbp2ozjeq37by2"
>         
> --- FAIL: TestAccPublicIpStatic_canLabelBe63 (6.83s)
> FAIL
> ```

Hi @katbyte thanks for your time. Test case has been updated and passed. Could you please take another look?

image

File: internal/services/network/public_ip_resource_test.go [hunk 1]
```diff
@@ -894,7 +895,7 @@ resource "azurerm_public_ip" "test" {
   sku               = "Basic"
   domain_name_label = "%s"
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, strings.ToLower(data.RandomStringOfLength(63)))
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, fmt.Sprintf("test%s", strings.ToLower(data.RandomStringOfLength(59))))
```
Comment (katbyte) [inline internal/services/network/public_ip_resource_test.go:898]:
```suggestion
`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, fmt.Sprintf("test%s", strings.ToLower(data.RandomStringOfLength(59)))) // prepend with test (4+59) to ensure the string starts with a letter
```

File: website/docs/r/public_ip.html.markdown [hunk 2]
```diff
@@ -60,6 +60,10 @@ The following arguments are supported:
 
 * `domain_name_label` - (Optional) Label for the Domain Name. Will be used to make up the FQDN. If a domain name label is specified, an A DNS record is created for the public IP in the Microsoft Azure DNS system.
 
+* `domain_name_label_scope` - (Optional) Scope for the domain name label. If a domain name label scope is specified, an A DNS record is created for the public IP in the Microsoft Azure DNS system with a hashed value includes in FQDN. Possible values are `NoReuse`, `ResourceGroupReuse`, `SubscriptionReuse` and `TenantReuse`.
+
+-> **Note** Once `domain_name_label_scope` is specified, it cannot be changed, changing it forces a new resource to be created.
```
Comment (katbyte) [inline website/docs/r/public_ip.html.markdown:65]:
```suggestion
* `domain_name_label_scope` - (Optional) Scope for the domain name label. If a domain name label scope is specified, an A DNS record is created for the public IP in the Microsoft Azure DNS system with a hashed value includes in FQDN. Possible values are `NoReuse`, `ResourceGroupReuse`, `SubscriptionReuse` and `TenantReuse`. Changing this forces a new Public IP to be created.
```

**Review by katbyte — APPROVED** [review]
LGTM 🍯

**Pull request merged**